### PR TITLE
[IMP] pos_restaurant: merge tables

### PR DESCRIPTION
**Pull request by vlst-odoo**
A common usecase in a restaurant scenario is merging tables. This commit
adds a new feature to merge tables in the floor plan view. The existing orders
are moved to the target table.

Upon order completion, the tables are automatically unmerged.

There is an option to unmerge the tables manually as well, from the
floor plan editor.

The state of the merged tables is persisted in the database.

Task: 3597485

https://github.com/odoo/enterprise/pull/52776

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 1]
```diff
@@ -244,9 +208,13 @@ export class FloorScreen extends Component {
         }
         newTable.floor_id = [this.activeFloor.id, ""];
         newTable.floor = this.activeFloor;
-        await this._save(newTable);
         this.activeTables.push(newTable);
         this.activeFloor.table_ids.push(newTable.id);
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:212]: move after the create

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 2]
```diff
@@ -477,45 +404,52 @@ export class FloorScreen extends Component {
...
         });
     }
-    async changeToCircle() {
-        await this.changeShape("round");
-    }
-    async changeToSquare() {
-        await this.changeShape("square");
+    stopOrderTransfer() {
+        this.pos.isTableToMerge = false;
+        this.pos.orderToTransfer = null;
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:421]: this should be in the 2nd commit

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 3]
```diff
@@ -477,45 +404,52 @@ export class FloorScreen extends Component {
...
     }
-    async setTableColor(color) {
-        const selectedTables = this.selectedTables;
-        selectedTables.forEach(async (selectedTable) => {
-            selectedTable.color = color;
-            await this._save(selectedTable);
-        });
+    async setColor(color) {
+        if (this.selectedTables.length > 0) {
+            this.selectedTables.forEach(async (selectedTable) => {
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:437]: remove async

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml [hunk 4]
```diff
@@ -1,52 +1,71 @@
...
-                        toggleEditMode.bind="toggleEditMode"
-            />
-            <div class="floor-selector d-flex text-center bg-100 fs-3 w-100 h-54">
+            <t t-set="editButtonClass" t-value="'text-center d-flex flex-column align-items-center btn btn-light text-uppercase'" />
+            <div t-if="pos.orderToTransfer" class="d-flex align-items-center justify-content-between px-3 border-bottom bg-view overflow-x-auto w-100">
+                <button t-attf-class="ms-auto {{editButtonClass}}" t-att-class="{active: !pos.isTableToMerge}" t-on-click="() => this.pos.isTableToMerge = false">
+                    <i class="fa fa-arrow-right" role="img" aria-label="Transfer" title="Transfer" />Transfer</button>
+                <button t-attf-class="{{editButtonClass}}" t-att-class="{active: pos.isTableToMerge}" t-on-click="() => this.pos.isTableToMerge = true">
+                    <i class="fa fa-link" role="img" aria-label="Merge" title="Merge" />Merge</button>
+                <button t-attf-class="ms-auto {{editButtonClass}}" t-on-click.stop="stopOrderTransfer">
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.xml:11]: all the stuff with merge has to go in the 2nd commit

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 5]
```diff
@@ -370,10 +299,9 @@ export class FloorScreen extends Component {
             this.pos.showScreen(order.get_screen_data().name);
         }
     }
-    async onSaveTable(table) {
-        if (this.pos.tables_by_id[table.id] && this.pos.tables_by_id[table.id].active) {
-            await this._save(table);
-        }
+    closeEditMode() {
+        this.pos.isEditMode = false;
+        this.state.selectedTableIds = [];
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:310]: remove the color selector at this point

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 6]
```diff
@@ -297,19 +243,8 @@ export class FloorScreen extends Component {
...
-        const tables = [];
-        this.state.selectedTableIds.forEach((id) => {
-            const table = this.pos.models["restaurant.table"].get(id);
-
-            if (table) {
-                tables.push(table);
-            }
-        });
-        return tables;
+        return this.state.selectedTableIds.map((id) => this.pos.models["restaurant.table"].get(id));
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:247]: is this get(id) correct ?

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 7]
```diff
@@ -132,7 +96,7 @@ export class FloorScreen extends Component {
                 newTable.position_v += 10;
             }
             delete newTable.id;
-            newTable.order_count = 0;
+            newTable.orders = 0;
```
Comment (vlst-odoo) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:99]: check this

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js (hunk 6 above)
Comment (davidmonnom) [reply addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:238]:
```suggestion
return this.pos.models["restaurant.table"].filter((t) => this.state.selectedTableIds.includes(t.id));
```

File: addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js [hunk 8]
```diff
@@ -487,45 +406,66 @@ export class FloorScreen extends Component {
...
+        if (this.selectedTables.length > 0) {
+            this.selectedTables.forEach((selectedTable) => {
+                selectedTable.color = color;
+            });
+            this.pos.updateTables(...this.selectedTables);
+        } else {
+            this.activeFloor.background_color = color;
+            this.pos.data.write("restaurant.floor", [this.activeFloor.id], {
+                background_color: color,
+            });
```
Comment (davidmonnom) [inline addons/pos_restaurant/static/src/app/floor_screen/floor_screen.js:441]: You don't need to use write from the models when isn't a related records (many2one, one2many etc..) you can directly assign the new value

```suggestion
           this.activeFloor.background_color = color,
```

**Review by caburj — APPROVED** [review]
robodoo r+

**Pull request closed without merging**